### ISSUE TITLE
Serializable isolation level not honored?

### ISSUE THREAD

**Issue by minollo** [godror/godror#295]
I have a simple scenario where I do:

- txn1 <- db.BeginTx(ctx, &sql.TxOptions{Isolation: sql.LevelReadCommitted, ReadOnly: false}
- txn2 <- db.BeginTx(ctx, &sql.TxOptions{Isolation: sql.LevelSerializable, ReadOnly: true}
- readSomeData(txn2): I get no results
- writeSomeData(txn1); I don't "commit" yet
- readSomeData(txn2): I get no results - good
- txn1.Commit()
- readSomeData(txn2): I get data written in txn1

I wouldn't expect that to happen; "serializable" should make sure txn2 won't see any external commits which happen after the transaction has started...
What am I missing?

**Comment by minollo** [comment]
It works as expected if in conn.BeginTx() I also execute...
  ALTER SESSION SET ISOLATION_LEVEL=SERIALIZABLE
...when the serializable level is requested; not sure why I would to do that - and whether that has some drawbacks.

**Comment by tgulacsi** [comment]
PTAL TestLevelSerializable in z_test.go at [7d9f260](https://github.com/godror/godror/commit/7d9f2609636f380bbbe3f36362a818f7bca95640) - it works as expected.

You can check what happens by setting DPI_DEBUG_LEVEL=92 and search for "SQL SET TRANSACTION ISOLATION LEVEL" in the output.

**Comment by tgulacsi** [comment]
READ ONLY or SERIALIZABLE ?
For an application standpoint, READ ONLY is more strict (does not allow modifications),
for DB level, SERIALIZABLE is the harder.

So the driver sets READ ONLY when you specify it, and SERIALIZABLE when you set {sql.LevelSerializable, ReadOnly: *false*}.

TL;DR; do not set ReadOnly when you don't mean it!

**Comment by minollo** [comment]
How does that test verifies it? You should have a tx2.commit() after the tx1.Exec() operation; but you don't - you eventually just rollback tx1 and tx2.
If you add tx1.Commit() after the "insert into" execution, I expect you will see the same issue I'm seeing.

**Comment by minollo** [comment]
In fact, if you add tx1.Commit() (ignore the lack of error checking):
...
	if _, err := tx1.ExecContext(ctx, "INSERT INTO "+tbl+" (id) VALUES (:1)", 3); err != nil {
		t.Fatal("insert1:", err)
	}
	if got := C(tx1); got != 1 {
		t.Errorf("tx1 has %d rows (instead of 1)", got)
	}
	tx1.Commit()
	if got := C(tx2); got != 0 {
		t.Errorf("tx2 has %d rows (instead of 0)", got)
	}
...

...you get:
=== RUN   TestLevelSerializable
    z_test.go:4799: 0xc00007e080: 0
    z_test.go:4799: 0xc00017e000: 0
    z_test.go:4799: 0xc00007e080: 1
    z_test.go:4799: 0xc00017e000: 1
    z_test.go:4814: tx2 has 1 rows (instead of 0)
--- FAIL: TestLevelSerializable (3.09s)

**Comment by tgulacsi** [comment]
The driver cannot change how Oracle works. It only can issue some commands to set up the transaction.

If you help what it shall do, I can change it.

**Comment by minollo** [comment]
I understand; someone from Oracle DB or DPI should probably look into this.
I can create a PR which shifts to use ALTER SESSION SET ISOLATION - and which corrects your test; then you can decide to do what you prefer with it.

**Comment by minollo** [comment]
https://github.com/godror/godror/pull/296

**Comment by minollo** [comment]
BTW, I may well be wrong, but I believe it makes sense to have both SERIALIZABLE and READ ONLY set at the same time - it covers the usecase where you are executing transactions which don't need to change data, but they also need to be insulated from commits happening concurrently through other sessions...

**Comment by tgulacsi** [comment]
Ok, I've reverted that commit, and I see the same suspicious behaviour: the READ ONLY transaction sees the later started, earlier commited rows.

I am definitely not a native English speaker, but https://docs.oracle.com/en/database/oracle/oracle-database/19/sqlrf/SET-TRANSACTION.html#GUID-F11E1E30-5871-48D1-8266-F80A1DF126A1 says 

    All subsequent queries in that transaction see only changes that were committed before the transaction began.

For me this means that a later started, earlier commited change should not be seen in a READ ONLY transaction.

This won't change when I change the order of SET TRANSACTION READ ONLY / SET TRANSACTION ISOLATION LEVEL SERIALIZABLE statements,
neither setting the RW transaction to SERIALIZABLE.

So, I'm out of my league here.
The only thing I know that Oracle's default READ COMMITED serialization level is the only really used and tested in Oracle;
SERIALIZABLE in Postgres does work, but has a lot of downsides (must restart transaction on conflict).

**Comment by minollo** [comment]
Yes, that documentation does say that READ ONLY is indeed a superset of SERIALIZABLE - sorry I missed that.

SERIALIZABLE in Oracle works perfectly well - I used it extensively with the oci8 mattn driver (based on OCI); and ALTER SESSION SET ISOLATION_LEVEL=SERIALIZABLE works perfectly well too, as I mentioned.

For some reason it looks like SET TRANSACTION ISOLATION LEVEL SERIALIZABLE does not work - and I have no idea why either; smells like a possible issue with ODPI, but I don't really know.
